### ISSUE TITLE
projections database add-on

### ISSUE THREAD

**Issue by hsharrison** [BurntSushi/nfldb#77]
Hey, thanks for this awesome library. I've been thinking of starting a new project, something similar to this but for keeping track of experts' projections.

I just thought I'd ask for advice, as I'd want it to integrate with `nfldb`--at least key off of game ID and player ID. But how tight is the question. Are things going to screw up if I had more tables to the database? And how hard would it be to leverage the query system as well? Appreciate any thoughts you might have.

As for how to make it available, I think the best option is a separate package with a dependency of `nfldb`. But if you think it would make more sense as a pull request, or even a fork, let me know.

**Comment by hsharrison** [comment]
@BurntSushi to keep track of the add-on version, would you recommend adding a value to `meta` or adding a new table?

**Comment by BurntSushi** [comment]
@hsharrison Unless you have a really good reason for doing so, I would very much try to stay away from modifying existing tables.

**Comment by hsharrison** [comment]
Makes sense, thanks.

**Comment by hsharrison** [comment]
The ERD is a big spaghetti mess but I think I've got all the major issues sorted bar one: handling games that aren't in nfldb yet. I'm at a bit of a loss here. The best plan I can think of is to use some placeholder value in `gsis_id` and have a separate script that scans for new games and updates the ids accordingly. It could be incorporated into an `nfldb-update` wrapper for convenience.

What do you think? Is there better option that I'm missing? This feels like an ugly hack.

**Comment by BurntSushi** [comment]
Could you outline the use case here? Can it just not add info for games that don't exist in the database? Generally speaking, games can be in nfldb as soon as the schedule for them is known. For example, when the season starts, an entry for every game in the season gets added as soon as the schedule is released. Similarly for playoffs---you just have to update the `nflgame` schedule.

**Comment by hsharrison** [comment]
Well, that solves the problem then!

I just saw the comment in `nfldb.update` about only inserting games into the database 15 minutes before kickoff (if I'm interpreting that correctly) so I guess the solution is to grab the IDs from `nflgame` instead of `nfldb`.

**Comment by BurntSushi** [comment]
Are you looking at the `games_schedule` function? Note the docstring. :-)

> Returns a list of GSIS identifiers corresponding to **games that
> have schedule data in the database** but don't have any drives or
> plays in the database. In the typical case, this corresponds to
> games that haven't started yet.
> 
> The list is sorted in the order in which the games will be played.

Unfortunately, the update script is rather complex. It has to be smart about what it updates, or else it will be really slow. e.g., It will bulk insert *schedule* data (which are just rows in the `game` table). But it has to avoid pinging NFL.com for drive/play data for *every* game, otherwise it would be pretty ridiculous. So it checks the start date/time of the game to see whether it should query for additional data.

And once a game is over, it never changes.

An interesting property of this approach is that it is *idempotent*. If you run `nfldb-update` on a completely updated database, it should make no changes. If you delete *any* row from the `game` table and re-run `nfldb-update`, that row and all of its drive/play data will get re-inserted automatically. (This is especially useful on the rare occasion that game data gets corrupted.)

`nfldb-update` is pretty smart about what it does, but it pays for that in complexity.

**Comment by hsharrison** [comment]
I was looking at [this comment](https://github.com/BurntSushi/nfldb/blob/master/nfldb/update.py#L263) just a few lines below that docstring. In short, games more than 15 minutes in the future are skipped. But now I see that's just for purposes of checking for more info rather than not inserting the game at all.

So I'm not sure why I didn't have the wild-card games in my database until I updated again just now. I thought I had updated yesterday or the day before, but I guess not.

Well, it looks like this was a false alarm. I'll just make sure `nfldb` is updated before inserting projections, and it should be able to find all the game IDs it needs. Idempotency will come in handy, then. :-)

**Comment by BurntSushi** [comment]
> So I'm not sure why I didn't have the wild-card games in my database until I updated again just now. I thought I had updated yesterday or the day before, but I guess not.

That's because I just updated `nflgame`. You can see this comment here: https://github.com/BurntSushi/nflgame/issues/106#issuecomment-68490377 --- that command will let you update the `nflgame` schedules without having to wait for me to do it (after which, you'll need to do a `pip install --upgrade nflgame`). Once you update the schedules in `nflgame`, the next run of `nfldb-update` will pick them up.

**Comment by hsharrison** [comment]
Ooooh, everything makes sense now. Awesome.

By the way, I've spent the last couple days looking through your code extensively (in order to mimic your approach wherever possible), and I have to say it's been way easier than I expected to figure out what's going on. So thanks for all the hard work you put into this, the final product is a joy to work with.

**Comment by BurntSushi** [comment]
No problem. :-) But you may not be saying that once you get to SQL generation. :-) This one is pretty crazy: https://github.com/BurntSushi/nfldb/blob/master/nfldb/sql.py#L177

**Comment by hsharrison** [comment]
Yeah, I almost added a line "but I haven't got the query system yet". :-)

**Comment by hsharrison** [comment]
Still making progress...

Any idea how I could handle DST "players"? The only option I have seems to be allowing the `player_id` column of the various projection tables to match with *either* `player.player_id` *or* `team.team_id`. Does Postgres allow this kind of dual foreign-key constraint? It sounds like something that will create headaches.

Alternatively, would you be willing to hardcode these rows into the `player` table? Or do you have any other ideas?

**Comment by BurntSushi** [comment]
I definitely do not want to be adding defensive teams as players in the `player` table. That seems very wrong.

In all the code I've written, I've handled "defense players" as a special case unfortunately. I haven't yet discovered a good abstraction over regular players and defensive team "players."

The other option is to have two projection tables. One for regular players and one for defensive teams. I'd probably try that.

**Comment by hsharrison** [comment]
> I definitely do not want to be adding defensive teams as players in the player table. That seems very wrong.

Fair enough.

> The other option is to have two projection tables. One for regular players and one for defensive teams. I'd probably try that.

Just curious, what's the rationale? A pro seems to be that I could then split the offensive statistics from the defensive categories. The con is that queries become a lot more complicated if one is interested in fantasy points across all positions.

Right now I'm leaning toward the intermediary table with a trigger to keep it updated when players are added. I think it will keep things easier, but I'm curious what your reason is to prefer splitting them.

**Comment by hsharrison** [comment]
Hmm, I got a second opinion for the option of splitting them, so I'm willing to accept my intuition here is wrong.

**Comment by BurntSushi** [comment]
I absolutely think you should go with your intuition. I'm not happy with any solution I've ever come up with---I'm just tossing out ideas. :-)

> Just curious, what's the rationale? A pro seems to be that I could then split the offensive statistics from the defensive categories. The con is that queries become a lot more complicated if one is interested in fantasy points across all "positions".

That second point is precisely the issue. If you split them, then running one query to grab everything is going to be difficult (so I agree with you there). It may be much easier to say, "If you want projections for individual players, then write this kind of SQL. And if you want projections for defensive teams, then write this kind of SQL." If you end up developing a library or a UI around this stuff, *that's* where you might be able to merge them under one common interface.

Your idea of using one id column to store *either* a player id or a team id is interesting. You're right to point out a foreign key problem. Maybe you just don't use a foreign key for that column?

Another option is for each of your projection tables to have two columns: `player_id` and `team_id` with the invariant that exactly one of them must always be `NULL`.

I really don't see a clear answer here. I'd suggest trying both and see which one works out better. :-)

**Comment by hsharrison** [comment]
Alright, thanks for the advice. Sounds like our thoughts on this are similar. I'll give..something...a shot.

**Comment by hsharrison** [comment]
Still thinking, but putting this link here for posterity as it touches upon all the options we just mentioned: http://stackoverflow.com/questions/441001/possible-to-do-a-mysql-foreign-key-to-one-of-two-possible-tables
